sled-agent/src/hardware/illumos/sysconf.rs

```diff
@@ -0,0 +1,40 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Access to sysconf-related info.
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("sysconf failed accessing {arg}: {e}")]
+    Sysconf { arg: String, e: std::io::Error },
+
+    #[error("Integer conversion error: {0}")]
+    Integer(#[from] std::num::TryFromIntError),
+}
+
+fn sysconf(argname: &str, arg: libc::c_int) -> Result<u64, Error> {
+    let r = unsafe { libc::sysconf(arg) };
+    if r == -1 {
+        return Err(Error::Sysconf {
+            arg: argname.to_string(),
+            e: std::io::Error::last_os_error(),
+        });
+    }
+    Ok(r.try_into()?)
+}
+
+/// Returns the number of physical processors on this sled.
+pub fn online_processor_count() -> Result<u32, Error> {
+    Ok(u32::try_from(sysconf(
```

Comment:
Separate to this, `sysconf` returns a long. In an ILP64 this is an i64. I realize we're unlikely to have more than a u32 of processors, but everywhere else you treat cpus as an `i64`.

Reply:
I thought we didn't want to enable negative values for CPUs? By using a u32 here, I validate the value from the sled. If I propagated an i64 all the way up to Nexus, it would be possible to use an "invalid" value for CPU count for much longer. Shouldn't we be throwing an error as early as we reasonably can?

Reply:
I'm honestly not sure what the right answer is to the fact unsigned types are for chumps in the db. Seems like probably best to just hope the constraint we have works and is never violated and so it can always serialize / de-serialize as a u32 and then what you're saying makes sense. Just whenever I see a constraint on that, my mind goes to what happens if that gets violated because of issue XXX and the wrong thing is on disk. As I said above, this probably isn't worth worrying about.